### PR TITLE
batch: Use array+join to avoid O(N^2) string concat.

### PR DESCRIPTION
**Pull request by nairb774**
String concatenation is a common source of useless work. This will cause
the individual string parts be collected up in the array (avoiding
repeated allocations, and memcopy) and only at the end, when all of the
parts have been accumulated, will the full string be built.

### CODE REVIEW

**Review by dmorina — APPROVED** [review]

**Comment by nairb774** [comment]
My JS is really weak - so please review carefully. No tests have been run if none are being run automatically.

File: lib/batch/index.js [hunk 1]
```diff
@@ -165,15 +165,16 @@ Batch.prototype.generateControl = function(cb) {
...
 	async.each(this._entries, function(entry, done) {
 		entry.generateString(function(string) {
-			result += string + utils.newLineChar();
+			result.push(string);
 			done();
 		});
 	}, function(err) {
-		cb(result);
+		result.push(''); // Force trailing new line.
+		cb(result.join(utils.newLineChar()));
```
Comment (dmorina) [inline lib/batch/index.js:177]: this should happen at the end of the func

Comment (nairb774) [reply lib/batch/index.js:177]: Not sure I follow, but I believe this code is running after all of the `async.each` code is being run, and the result is being returned back (via a callback) to the caller.

**Review by dmorina — APPROVED** [review]

File: lib/batch/index.js (hunk 1 above)
Comment (crotger) [reply lib/batch/index.js:177]: normally this callback function would want to handle the error param defined on :175 but it doesn't actually seem possible for it to be non-null here

File: lib/batch/index.js [hunk 2]
```diff
@@ -165,15 +165,16 @@ Batch.prototype.generateControl = function(cb) {
 };
 
 Batch.prototype.generateEntries = function(cb) {
-	var result = '';
+	var result = [];
 
 	async.each(this._entries, function(entry, done) {
 		entry.generateString(function(string) {
-			result += string + utils.newLineChar();
+			result.push(string);
```
Comment (crotger) [inline lib/batch/index.js:172]: so just so I'm clear, the logic is: string concat is O(N) in the size of the string, each loop grows the string by N, which means overall we are doing N^2?

**Review by crotger — APPROVED** [review]
Code looks fine to me. You could test by changing out the pinned version locally in digit-vault for this feature branch, then running the fixture tests for NACHA files.

File: lib/batch/index.js (hunk 2 above)
Comment (nairb774) [reply lib/batch/index.js:172]: Each += causes the prior string to be copied to a new string. This causes the runtime to grow O(N^2) in the number of bytes. Strings and + are never a good idea in large quantities. This is really only a small improvement overall. I focused on this change by reading the description of how the system was behaving and focusing on the code that could contribute to that. There are other suspect +=s and +s in there that likely could be improved - just trying to take the edge off.

File: lib/batch/index.js [hunk 3]
```diff
@@ -165,15 +165,16 @@ Batch.prototype.generateControl = function(cb) {
 };
 
 Batch.prototype.generateEntries = function(cb) {
-	var result = '';
+	var result = [];
```
Comment (ajones) [inline lib/batch/index.js:168]:
```
Appending - Amortized O(1) (sometimes resizing the hashtable is required; usually, only insertion is required)
```
You could have guarded from this with `new Array(this._entries.length)`

**Pull request merged**